airbyte-integrations/connectors/source-amazon-seller-partner/source_amazon_seller_partner/streams.py

```diff
@@ -102,79 +129,286 @@ def get_updated_state(self, current_stream_state: MutableMapping[str, Any], late
             return {self.cursor_field: max(latest_benchmark, current_stream_state[self.cursor_field])}
         return {self.cursor_field: latest_benchmark}
 
+
+class ReportsAmazonSPStream(Stream, ABC):
+    """
+    API docs: https://github.com/amzn/selling-partner-api-docs/blob/main/references/reports-api/reports_2020-09-04.md
+    API model: https://github.com/amzn/selling-partner-api-models/blob/main/models/reports-api-model/reports_2020-09-04.json
+
+    Report streams are intended to work as following:
+        - create a new report;
+        - retrieve the report;
+        - retry the retrieval if the report is still not fully processed;
+        - retrieve the report document (if report processing status is `DONE`);
+        - decrypt the report document (if report processing status is `DONE`);
+        - yield the report document (if report processing status is `DONE`)
+        or the report json (if report processing status is **not** `DONE`)
```

Comment:
What is the report json? if the report is not done then this is a failure right? meaning Airbyte is not able to pull the report? shouldn't we throw an exception if that is the case?

Reply:
I guess my guiding principle here is: if Airbyte is not able to pull the data for the user then the sync should not succeed. Because in that situation the only way the user will see the issue is downstream.

Reply:
Oh, I've got your point there. For now we're yielding either the report document if the one exists or the report body if not. I though it may be useful as user may notice what's to change in their setup (e.g. if report status is `CANCELLED` then it's either cancelled manually or there is no data to generate the report).

So do you think it's a good point to raise the exception on `CANCELLED` report status? Because it's not an issue at any side, just no data to return.

Reply:
Also if we're going to return only document content, then we'll not able to enable any acceptance tests except `spec`, `connection` and `discovery`. Because all of the streams are empty.

Reply:
How can a report get into the `CANCELLED` state? does it have to be a manual intervention or is there some "acceptable" way for a report to become CANCELLED that doesn't indicate an error status? If the former then we should error

Let's mark those streams as empty in acceptance tests

Reply:
`CANCELLED` state means that either there is no data to generate the report or user cancelled the generation manually. Neither of it is error status.
From the docs: _The report was cancelled. There are two ways a report can be cancelled: an explicit cancellation request before the report starts processing, or an automatic cancellation if there is no data to return._
For the error case there is `FATAL` state.